### DISCUSSION TITLE
Docker Compose Nextjs Nodejs

### DISCUSSION THREAD

**Discussion by Mohamed-Abdulla** [vercel/next.js#58893] [App Router]
Hey iam new to docker, iam trying to dockerize nextjs14 and node js. 
when i docker compose build and up. i got this error
2023-11-25 09:18:40 Node.js v18.18.2
2023-11-25 09:18:43 node:internal/modules/cjs/loader:1080
2023-11-25 09:18:43   throw err;
2023-11-25 09:18:43   ^
2023-11-25 09:18:43 
2023-11-25 09:18:43 Error: Cannot find module '/app/server.js'
2023-11-25 09:18:43     at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
2023-11-25 09:18:43     at Module._load (node:internal/modules/cjs/loader:922:27)
2023-11-25 09:18:43     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:86:12)
2023-11-25 09:18:43     at node:internal/main/run_main_module:23:47 {
2023-11-25 09:18:43   code: 'MODULE_NOT_FOUND',
2023-11-25 09:18:43   requireStack: []
2023-11-25 09:18:43 }
2023-11-25 09:18:43

my nextconfig.js
const nextConfig = {
  output: "standalone",
};
I'll attach my docker files  and appreciate help. thanks in advance🤝
-----
This is my nextjs Dockerfile

FROM node:18-alpine AS base

FROM base AS builder

WORKDIR /app

COPY package.json yarn.lock* package-lock.json* pnpm-lock.yaml* ./

RUN \
  if [ -f yarn.lock ]; then yarn --frozen-lockfile; \
  elif [ -f package-lock.json ]; then npm ci; \
  elif [ -f pnpm-lock.yaml ]; then yarn global add pnpm && pnpm i; \
  else echo "Warning: Lockfile not found. It is recommended to commit lockfiles to version control." && yarn install; \
  fi

COPY . .
COPY public ./public
COPY next.config.js .
COPY tsconfig.json .

ARG NEXT_PUBLIC_URL
ARG NEXT_PUBLIC_REGENERATE_URL
ARG NEXT_PUBLIC_ENV
ARG NEXT_PUBLIC_CREATED_BY
ARG NEXT_PUBLIC_PAGE_SIZE
ARG NEXT_PUBLIC_API_SEARCH_EMAIL_MAIL_CAMPAIGN
ARG NEXT_PUBLIC_API_SEARCH_EMAIL_CONTACT
ARG NEXT_PUBLIC_API_SEARCH_EMAIL_COMPANY

ENV NEXT_PUBLIC_URL=${NEXT_PUBLIC_URL}
ENV NEXT_PUBLIC_REGENERATE_URL=${NEXT_PUBLIC_REGENERATE_URL}
ENV NEXT_PUBLIC_ENV=${NEXT_PUBLIC_ENV}
ENV NEXT_PUBLIC_CREATED_BY=${NEXT_PUBLIC_CREATED_BY}
ENV NEXT_PUBLIC_PAGE_SIZE=${NEXT_PUBLIC_PAGE_SIZE}
ENV NEXT_PUBLIC_API_SEARCH_EMAIL_MAIL_CAMPAIGN=${NEXT_PUBLIC_API_SEARCH_EMAIL_MAIL_CAMPAIGN}
ENV NEXT_PUBLIC_API_SEARCH_EMAIL_CONTACT=${NEXT_PUBLIC_API_SEARCH_EMAIL_CONTACT}
ENV NEXT_PUBLIC_API_SEARCH_EMAIL_COMPANY=${NEXT_PUBLIC_API_SEARCH_EMAIL_COMPANY}

RUN \
  if [ -f yarn.lock ]; then yarn build; \
  elif [ -f package-lock.json ]; then npm run build; \
  elif [ -f pnpm-lock.yaml ]; then pnpm build; \
  else yarn build; \
  fi

FROM base AS runner

WORKDIR /app

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs
USER nextjs

COPY --from=builder /app/public ./public
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

COPY .env .env
ARG NEXT_PUBLIC_URL
ARG NEXT_PUBLIC_REGENERATE_URL
ARG NEXT_PUBLIC_ENV
ARG NEXT_PUBLIC_CREATED_BY
ARG NEXT_PUBLIC_PAGE_SIZE
ARG NEXT_PUBLIC_API_SEARCH_EMAIL_MAIL_CAMPAIGN
ARG NEXT_PUBLIC_API_SEARCH_EMAIL_CONTACT
ARG NEXT_PUBLIC_API_SEARCH_EMAIL_COMPANY

ENV NEXT_PUBLIC_URL=${NEXT_PUBLIC_URL}
ENV NEXT_PUBLIC_REGENERATE_URL=${NEXT_PUBLIC_REGENERATE_URL}
ENV NEXT_PUBLIC_ENV=${NEXT_PUBLIC_ENV}
ENV NEXT_PUBLIC_CREATED_BY=${NEXT_PUBLIC_CREATED_BY}
ENV NEXT_PUBLIC_PAGE_SIZE=${NEXT_PUBLIC_PAGE_SIZE}
ENV NEXT_PUBLIC_API_SEARCH_EMAIL_MAIL_CAMPAIGN=${NEXT_PUBLIC_API_SEARCH_EMAIL_MAIL_CAMPAIGN}
ENV NEXT_PUBLIC_API_SEARCH_EMAIL_CONTACT=${NEXT_PUBLIC_API_SEARCH_EMAIL_CONTACT}
ENV NEXT_PUBLIC_API_SEARCH_EMAIL_COMPANY=${NEXT_PUBLIC_API_SEARCH_EMAIL_COMPANY}

CMD ["node", "server.js"]

----------------------

this is my nodejs Docker file

FROM node:18-alpine

WORKDIR /app

COPY package*.json ./

RUN npm install

COPY . .

EXPOSE 5000

CMD [ "npm", "start" ]
----------------

this is my DOcker compose file

version: "3.5"

services:
  ba_tool_backend:
    container_name: ba-tool-backend
    build:
      context: ./oracle-api
    ports:
      - "5000:5000"
    environment:
      - ENV=production

    volumes:
      - ./oracle-api:/app
    networks:
      - ba_tool_network

  ba_tool_frontend:
    container_name: ba-tool-frontend
    build:
      context: ./colakin-ba-tool-ui
      args:
        - NEXT_PUBLIC_URL=http://localhost:5000/api/v1/
        - NEXT_PUBLIC_REGENERATE_URL=http://127.0.0.1:8000/regenerate/
        - NEXT_PUBLIC_ENV=production
        - NEXT_PUBLIC_CREATED_BY=Abdulla
        - NEXT_PUBLIC_PAGE_SIZE=20
        - NEXT_PUBLIC_API_SEARCH_EMAIL_MAIL_CAMPAIGN=http://localhost:5000/api/v1/mail_campaign/search?email=
        - NEXT_PUBLIC_API_SEARCH_EMAIL_CONTACT=http://localhost:5000/api/v1/contacts/search?email=
        - NEXT_PUBLIC_API_SEARCH_EMAIL_COMPANY=http://localhost:5000/api/v1/companies/search?name=
    restart: always
    ports:
      - "3000:3000"
    # env_file: .env
    volumes:
      - ./colakin-ba-tool-ui:/app
    depends_on:
      - ba_tool_backend
    networks:
      - ba_tool_network
    environment:
      - NEXT_PUBLIC_URL=http://localhost:5000/api/v1/
      - NEXT_PUBLIC_REGENERATE_URL=http://127.0.0.1:8000/regenerate/
      - NEXT_PUBLIC_ENV=production
      - NEXT_PUBLIC_CREATED_BY=Abdulla
      - NEXT_PUBLIC_PAGE_SIZE=20
      - NEXT_PUBLIC_API_SEARCH_EMAIL_MAIL_CAMPAIGN=http://localhost:5000/api/v1/mail_campaign/search?email=
      - NEXT_PUBLIC_API_SEARCH_EMAIL_CONTACT=http://localhost:5000/api/v1/contacts/search?email=
      - NEXT_PUBLIC_API_SEARCH_EMAIL_COMPANY=http://localhost:5000/api/v1/companies/search?name=

networks:
  ba_tool_network:
    external: true

**Comment by vahidirp** [comment]
Hi,
I think this example should solve your problem.

docker-compose.yml file content : 
```
version: '3'
services:
  nextjs:
    build:
      context: ./nextjs
    ports:
      - "3000:3000"
    volumes:
      - ./nextjs:/app
    command: sh -c "yarn install && yarn dev"

  nodejs:
    build:
      context: ./nodejs
    volumes:
      - ./nodejs:/app
    command: sh -c "yarn install && node server.js"
```

and Dockerfiles are in two directory

/nextjs/Dockerfile with this content maybe : 

``` 
# nextjs/Dockerfile
FROM node:14-alpine

WORKDIR /app

COPY package.json yarn.lock ./

RUN yarn install

COPY . .

CMD ["yarn", "dev"]
```

/nodejs/Dockerfile with this content maybe : 
```
# nodejs/Dockerfile
FROM node:14-alpine

WORKDIR /app

COPY package.json yarn.lock ./

RUN yarn install

COPY . .

CMD ["node", "server.js"]
```
